src/LevelAgreement.php

```diff
@@ -875,18 +888,32 @@ public function computeExecutionDate($start_date, $levels_id, $additional_delay
 
             if ($level->getFromDB($levels_id)) { // level exists
                 if ($level->fields[$fk] == $this->fields['id']) { // correct level
-                    $work_in_days = ($this->fields['definition_time'] == 'day' || $this->fields['definition_time'] == 'month');
                     $delay        = $this->getTime();
 
                    // Based on a calendar
                     if ($this->fields['calendars_id'] > 0) {
                         $cal = new Calendar();
                         if ($cal->getFromDB($this->fields['calendars_id']) && $cal->hasAWorkingDay()) {
-                            return $cal->computeEndDate(
+                            // Take SLA into account
+                            $date_with_sla = $cal->computeEndDate(
                                 $start_date,
                                 $delay,
-                                $level->fields['execution_time'] + $additional_delay,
-                                $work_in_days
+                                0,
+                                $this->shouldUseWorkInDayMode()
+                            );
+
+                            // Take current SLA escalation level into account
+                            $date_with_sla_and_escalation_level = $cal->computeEndDate(
+                                $date_with_sla,
+                                $level->fields['execution_time'],
+                                0,
+                                $level->shouldUseWorkInDayMode()
+                            );
+
+                            // Take waiting duration time which give us the final date
+                            return $date_with_sla_and_escalation_level = $cal->computeEndDate(
```

Comment:
Are you sure the waiting time has to be applied after the execution delay?

Reply:
Unless I'm missing something it doesn't really matter when it is added, results would be the same with the 3 dates calculated in any orders ?

Reply:
You are right, after further thinking and re-reading my tests case, the waiting time should in fact be calculated before the escalation level.